lib/valueflow.cpp

```diff
@@ -2501,49 +2506,64 @@ static void valueFlowAfterCondition(TokenList *tokenlist, SymbolDatabase* symbol
+                bool check_if = false;
+                bool check_else = false;
                 if (Token::Match(tok, "==|>=|<=|!|>|<"))
-                    codeblock = 1;
-                else if (Token::Match(tok, "%name%|!="))
-                    codeblock = 2;
+                    check_if = true;
+                if (Token::Match(tok, "%name%|!=|>|<"))
+                    check_else = true;
 
                 // determine startToken based on codeblock
-                if (codeblock > 0) {
+                if (check_if || check_else) {
                     // if astParent is "!" we need to invert codeblock
                     const Token *parent = tok->astParent();
                     while (parent && parent->str() == "&&")
                         parent = parent->astParent();
-                    if (parent && parent->str() == "!")
-                        codeblock = (codeblock == 1) ? 2 : 1;
+                    if (parent && parent->str() == "!") {
+                        check_if = !check_if;
+                        check_else = !check_else;
+                    }
 
                     // convert codeblock to a startToken
-                    if (codeblock == 1 && Token::simpleMatch(top->link(), ") {"))
-                        startToken = top->link()->next();
-                    else if (Token::simpleMatch(top->link()->linkAt(1), "} else {"))
-                        startToken = top->link()->linkAt(1)->tokAt(2);
-                }
-
-                if (startToken) {
-                    if (values.size() == 1U && Token::Match(tok, "==|!")) {
-                        const Token *parent = tok->astParent();
-                        while (parent && parent->str() == "&&")
-                            parent = parent->astParent();
-                        if (parent && parent->str() == "(")
-                            values.front().setKnown();
-                    }
-                    if (!valueFlowForward(startToken->next(), startToken->link(), var, varid, values, true, false, tokenlist, errorLogger, settings))
-                        continue;
-                    values.front().setPossible();
-                    if (isVariableChanged(startToken, startToken->link(), varid, var->isGlobal(), settings)) {
-                        // TODO: The endToken should not be startToken->link() in the valueFlowForward call
-                        if (settings->debugwarnings)
-                            bailout(tokenlist, errorLogger, startToken->link(), "valueFlowAfterCondition: " + var->name() + " is changed in conditional block");
-                        continue;
+                    if (check_if && Token::simpleMatch(top->link(), ") {"))
+                        startTokens[0] = top->link()->next();
+                    if (check_else && Token::simpleMatch(top->link()->linkAt(1), "} else {"))
+                        startTokens[1] = top->link()->linkAt(1)->tokAt(2);
+                }
+
+                int bail = 0;
```

Comment:
this "bail" variable is questionable. I don't see that it has any effect if it's 1 or 0 before the loop.

Reply:
I guess its a little confusing, so maybe it can be written in a better way. If we have `x == 1`, then `check_if` is true and we will do `continue` when valueflow is false. If we have `x != 1`, then `check_else` is true and we will do `continue` when valueflow is false. However, for `x > 1` or `x < 1` then both `check_if` and `check_else` is true, so I only want to do `continue` when valueflow fails for both branches.

Reply:
there are two possible actions for bail in the loop:
 * it is incremented twice.
 * it is set to -1
so the `if (bail < 2)` can be rewritten as `if (bail == -1)` as far as I see.

Reply:
Its only incremented when `valueFlowForward` is true.